[DOC] Alignment IO (#808)
seqan/seqan3: doc/tutorial/alignment_file/index.md

# Alignment Input and Output in SeqAn3 {#tutorial_alignment_file}

***Learning Objective:***

<b>Learning Objective:</b> <br/>
You will get an overview of how to read and write alignment files.
This tutorial is a walk-through with links into the API documentation and also meant as a source for copy-and-paste code.

\tutorial_head{Medium, 60 min, \ref setup\, \ref tutorial_alphabets\, \ref tutorial_sequence_file,}

[TOC]

# Introduction

Alignment files are used to store pairwise alignments between two (biological) sequences.
Common file formats are the Sequence Alignment/Map format (SAM) and BLAST output format.
Next to the alignment, those formats store additional information like the start positions or mapping qualities.
Alignment files are a little more complex than sequence files but the basic design is the same.
If you are new to SeqAn3, we strongly recommend to do the [Sequence File Tutorial](#tutorial_alignment_file) first.

> 

Reviewer: Maybe give a link to the spec?

Author: I added a link to the spec in the SAM format section

# Alignment file formats

## SAM format

SAM stands for Sequence Alignment/Map format. It is a TAB-delimited text format consisting of a header
section, which is optional, and an alignment section.

Here is an example of a SAM file:

```
@HD VN:1.6 SO:coordinate
@SQ SN:ref LN:45
r001   99 ref  7 30 8M2I4M1D3M = 37  39 TTAGATAAAGGATACTG *
r003    0 ref  9 30 5S6M       *  0   0 GCCTAAGCTAA       * SA:Z:ref,29,-,6H5M,17,0;
r004    0 ref 16 30 6M14N5M    *  0   0 ATAGCTTCAGC       *
r003 2064 ref 29 17 6H5M       *  0   0 TAGGC             * SA:Z:ref,9,+,5S6M,30,1;
r001  147 ref 37 30 9M         =  7 -39 CAGCGGCAT         * NM:i:1
```

> The following table summarized the columns in a SAM file:

Reviewer: ```suggestion
The following table summarises the columns of a SAM file:
```

| #  | SAM Column ID |  Description                                      |
|:--:|:--------------|:--------------------------------------------------|
| 1  | QNAME         | Query template NAME                               |
| 2  | FLAG          | bitwise FLAG                                      |
| 3  | RNAME         | Reference sequence NAME                           |
| 4  | POS           | 1-based leftmost mapping POSition                 |
| 5  | MAPQ          | MAPping Quality                                   |
| 6  | CIGAR         | CIGAR string                                      |
| 7  | RNEXT         | Reference name of the mate/next read              |
| 8  | PNEXT         | Position of the mate/next read                    |
| 9  | TLEN          | observed Template LENgth                          |
| 10 | SEQ           | segment SEQuence                                  |
| 11 | QUAL          | ASCII of Phred-scaled base QUALity+33             |

> 

Reviewer: double newline

If you want to read more about the SAM format, take a look at the
[official specifications](https://samtools.github.io/hts-specs/SAMv1.pdf).

## BAM format

BAM is the binary format version of SAM. It provides the same data as the SAM format
with the only difference that **the header is mandatory**.

# Alignment file fields

> The Alignment file abstraction supports writing following fields:

Reviewer: ```suggestion
The Alignment file abstraction supports writing the following fields:
```

The SAM format is the common output format of read mappers where you align short read sequence
to one or more large reference sequences.
In fact, the SAM format stores those alignment information only partially:
It **does not store the reference sequence** but only the read sequence and a *CIGAR* string
representing the alignment based on the read.

Take this SAM record as an example:

```
r003   73   ref   3   17   1M1D4M   *  0   0  TAGGC   *
```

The record gives you the following information:
A read with name `r003` has been mapped to a reference with name `ref` at position `3`
(in the reference, counting from 1) with a quality of `17` (phred scaled).
The flag has a value of `73` which indicates that the read is paired, the first in pair, but the mate is unmapped
(see [this website](https://broadinstitute.github.io/picard/explain-flags.html) for a nice explanation of SAM flags).
Fields set to `0` or `*` are defaulted and contain no information.

The cigar string is `1M1D4M` which represents the following alignment:

```
      1 2 3 4 5 6 7 8 9 ...
ref   N N N N N N N N N ...
read      T - A G G C
```

where the reference sequence is not known (represented by `N`).

> You will learn in the next section how to handle additional reference sequence information.

Reviewer: Maybe give a link to some external resource on cigar strings?

Author: I linked the  [SAM specifications](https://samtools.github.io/hts-specs/SAMv1.pdf) and the [SAMtools paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2723002/).
Since all other articles I found are either a copy of the specs or questions and answers in forums.

1. field::SEQ
2. field::ID
3. field::OFFSET
4. field::REF_SEQ
5. field::REF_ID
6. field::REF_OFFSET
7. field::ALIGNMENT
8. field::MAPQ
9. field::FLAG
10. field::QUAL
11. field::MATE
12. field::TAGS
13. field::EVALUE
14. field::BIT_SCORE

There is an additional field called seqan3::field::HEADER_PTR.
It is used to transfer header information from seqan3::alignment_file_input to seqan3::alignment_file_output,
but you needn't deal with this field manually.

Note that some of the fields are specific to the SAM format, while some are specific to BLAST.
To make things clearer, here is the table of SAM columns connected to the corresponding alignment file field:

| #  | SAM Column ID |  FIELD name                                       |
|:--:|:--------------|:--------------------------------------------------|
| 1  | QNAME         | seqan3::field::ID                                 |
| 2  | FLAG          | seqan3::field::FLAG                               |
| 3  | RNAME         | seqan3::field::REF_ID                             |
| 4  | POS           | seqan3::field::REF_OFFSET                         |
| 5  | MAPQ          | seqan3::field::MAPQ                               |

> | 6  | CIGAR         | implicilty stored in seqan3::field::ALIGNMENT     |

Reviewer 2: ```suggestion
| 6  | CIGAR         | implicitly stored in seqan3::field::ALIGNMENT     |
```

\snippet doc/tutorial/alignment_file/alignment_file_snippets.cpp read_custom_fields

\attention The order in which you specify the selected fields determines the order of elements in the seqan3::record.

In the example above we only select the id, sequence and flag information
so the seqan3::record object has three tuple elements that are decomposed using structural bindings.

Note that this is possible for all seqan3 file objects.

\assignment{Exercise: Accumulating mapping qualities}

Let's assume we want to compute the average mapping quality of a SAM file.

For this purpose, write a small program that
    * only reads the mapping quality (field::MAPQ) out of a SAM file and
    * computes the average of all qualities.

Use the following file to test your program:

\snippet doc/tutorial/alignment_file/alignment_file_solution1.cpp sam_file

It should output:
```
Average: 27.4
```

\endassignment
\solution

\snippet doc/tutorial/alignment_file/alignment_file_solution1.cpp solution

\endsolution

# Alignment representation in the SAM format

In SeqAn3 we represent an alignment as a tuple of two *aligned_sequences*,
as you have probably learned by now from the alignment tutorial.

> The SAM format is the common output format of read mappers where you align short read sequence

Reviewer 2: ```suggestion
The SAM format is the common output format of read mappers where you align short read sequences
```

to one or more large reference sequences.
In fact, the SAM format stores those alignment information only partially:
It **does not store the reference sequence** but only the read sequence and a *CIGAR* string
representing the alignment based on the read.

Take this SAM record as an example:

```
r003   73   ref   3   17   1M1D4M   *  0   0  TAGGC   *
```

The record gives you the following information:
A read with name `r003` has been mapped to a reference with name `ref` at position `3`

> (in the reference, counting from 1) with a quality of `17` (phred scaled).

Reviewer 2: ```suggestion
(in the reference, counting from 1) with a quality of `17` (Phred scaled).
```

I think it's always capital P?

take a look at the [SAM specifications](https://samtools.github.io/hts-specs/SAMv1.pdf)
or the [SAMtools paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2723002/).

## Completing reference information

In SeqAn3 the conversion from a CIGAR string to an alignment (two *aligned_sequences*) is done automatically for you.
You can just pass the alignment object to the alignment file:

\snippet doc/tutorial/alignment_file/alignment_file_snippets.cpp alignments_without_ref

In the example above, you can only safely access the aligned read.

\attention The **unknown** aligned reference sequence at the first position in the alignment tuple cannot be accessed
           (e.g. via the `operator[]`). It is represented by a dummy type that throws on access.

Although the SAM format does not handle reference sequence information,
you can provide these information to the seqan3::alignment_file_input which automatically fills the alignment object.
You can pass reference ids and reference sequences as additional constructor parameters:

\snippet doc/tutorial/alignment_file/alignment_file_snippets.cpp alignments_with_ref

\assignment{Exercise: Combining sequence and alignment files}

> Read in the following reference sequence fasta file (see the sequence file tutorial if you need a remainder):

Reviewer 2: ```suggestion
Read in the following reference sequence FASTA file (see the sequence file tutorial if you need a remainder):
```

\snippet doc/tutorial/alignment_file/alignment_file_solution2.cpp ref_file

Then read in the following SAM file while providing the reference sequence information.
Only read in the id, reference id, mapping quality, and alignment.

\snippet doc/tutorial/alignment_file/alignment_file_solution2.cpp sam_file

With those information do the following:
  * Filter the alignment records and only take those with a mapping quality >= 30.

>     (Take a look at the sequence file tutorial "fun with ranges" if you need a remainder)

Reviewer 2: Can you put a link here? (At least to the seq file tut)

>     (Take a look at the sequence file tutorial "fun with ranges" if you need a remainder)

Reviewer 2: ```suggestion
    (Take a look at the sequence file tutorial "fun with ranges" if you need a reminder)
```
